### PR TITLE
Add test for issue when excluding a nested dir from a higher level gitignore file

### PR DESCRIPTION
**Pull request by jakemac53**
Repro test for https://github.com/dart-lang/pub/issues/4300

I don't see a good mechanism for skipping individual parts of a test so this can just land after the issue is fixed (or close this and write different tests)

### CODE REVIEW

File: test/validator/gitignore_test.dart [hunk 1]
```diff
@@ -51,6 +51,13 @@ void main() {
       ]),
       exit_codes.DATA,
     );
+
+    // Check that you can unignore certain files too.
+    await d.dir('myapp', [
+      d.file('.gitignore', '*.txt\n!foo.txt'),
```
Comment (sigurdm) [inline test/validator/gitignore_test.dart:57]: I believe this is tested well enough in the general ignore-rules tests. Eg. here: https://github.com/dart-lang/pub/blob/e1fbda73d1ac597474b82882ee0bf6ecea5df108/test/ignore_test.dart#L344

File: test/validator/gitignore_test.dart [hunk 2]
```diff
@@ -121,6 +128,17 @@ void main() {
       exit_codes.DATA,
       workingDirectory: packageRoot,
     );
+
+    // Check that you can unignore certain dirs from a higher level gitignore.
+    await d.dir('reporoot', [
+      d.file('.gitignore', '*.txt\n!myapp/'),
```
Comment (sigurdm) [inline test/validator/gitignore_test.dart:134]: I don't think this does whatever you think it does.

`myapp/` is not ignored here, so it cannot meaningfully be negated/unignored.

We could test that `"/*\n!myapp/"` leads to no warnings. That is already the case.
```suggestion
      d.file('.gitignore', '/*\n!myapp/'),
```

**Comment by sigurdm** [comment]
As mentioned in the issue I believe you are "holding it wrong".

I think we implement the correct rules here. They are just somewhat obscure.

**Comment by jakemac53** [comment]
Sounds good, that behavior seems bizarre to me but as long as we are consistent with git, that is what matters :)

**Pull request closed without merging**